examples/mpi/mpi_overlap_bcast_mps.py

```diff
@@ -0,0 +1,173 @@
+      functionalities from our extension (and the backend itself) could be used
+      in a similar script.
+    - The reason this is the fastest approach is that we want to do as much work as
+      possible outside of the loop that computes the inner products: this loop iterates
+      ``O(n_circs^2)`` times, but we only need to contract ``O(n_circ)`` MPS objects.
+"""
+
+import sys
+from random import random
+
+from cupy.cuda.runtime import getDeviceCount
+from mpi4py import MPI
+
+from pytket.circuit import Circuit, fresh_symbol
+
+from pytket.extensions.cutensornet.mps import simulate, ContractionAlg
+
+# Parameters
+if len(sys.argv) < 3:
+    print(f"You need call this script as {sys.argv[0]} <n_qubits> <n_circs>")
+n_qubits = int(sys.argv[1])
+n_circs = int(sys.argv[2])
+
+root = 0
+comm = MPI.COMM_WORLD
+
+rank, n_procs = comm.Get_rank(), comm.Get_size()
+# Assign GPUs uniformly to processes
+device_id = rank % getDeviceCount()
+
+time_start = MPI.Wtime()
+mps_list = []
+
+if n_circs % n_procs != 0:
+    raise RuntimeError(
+        "Current version requires that n_circs is a multiple of n_procs."
+    )
+
+if rank == root:
+    print("\nGenerating the circuits.")
+    time0 = MPI.Wtime()
+
+# Generate the list of circuits in parallel
+circs_per_proc = n_circs // n_procs
+this_proc_circs = []
+
+# Generate the symbolic circuit
+sym_circ = Circuit(n_qubits)
+even_qs = sym_circ.qubits[0::2]
+odd_qs = sym_circ.qubits[1::2]
+
+for q0, q1 in zip(even_qs, odd_qs):
+    sym_circ.TK2(fresh_symbol(), fresh_symbol(), fresh_symbol(), q0, q1)
+for q in sym_circ.qubits:
+    sym_circ.H(q)
+for q0, q1 in zip(even_qs[1:], odd_qs):
+    sym_circ.TK2(fresh_symbol(), fresh_symbol(), fresh_symbol(), q0, q1)
+free_symbols = sym_circ.free_symbols()
+
+# Create each of the circuits
+for _ in range(circs_per_proc):
+    symbol_map = {symbol: random() for symbol in free_symbols}
+    my_circ = sym_circ.copy()
+    my_circ.symbol_substitution(symbol_map)
+    this_proc_circs.append(my_circ)
+
+if rank == root:
+    time1 = MPI.Wtime()
+    print(f"Circuit list generated. Time taken: {time1-time0} seconds.\n")
+    print("Contracting the MPS of the circuits.")
+    sys.stdout.flush()
+    time0 = MPI.Wtime()
+
+# Contract the MPS of each of the circuits in this process
+this_proc_mps = []
+for circ in this_proc_circs:
+    this_proc_mps.append(simulate(circ, ContractionAlg.MPSxGate, device_id=device_id))
+
+if rank == root:
+    time1 = MPI.Wtime()
+    print(f"All MPS contracted. Time taken: {time1-time0} seconds.\n")
+    print("Broadcasting the MPS of the circuits.")
+    sys.stdout.flush()
+
+# Broadcast the list of MPS
+time0 = MPI.Wtime()
+for proc_i in range(n_procs):
+    mps_list += comm.bcast(this_proc_mps, proc_i)
+# Change device ID
+# TODO: I don't think this is moving mem between GPUs on same node. Look into NCCL.
+for mps in mps_list:
+    mps._device_id = device_id
+    mps.init_cutensornet()
```

Comment:
Hmm why do you need to do that? I thought you only are supposed to use it as a context manager?

Reply:
Yeah, I'm being a bit sneaky here and I just realised I never destroy the handle... 

The reason is that initialising cutTN takes a considerable amount of time, so we don't want to do it everytime we compute `vdot` in the loop below. So, in this case, using the context manager approach is not ideal. 

In fact, all MPS here could be using the same handle to the cuTN library, and this kind of situation I did not consider when designing `init_cutensornet`. Now that I think about it, I think it'd be much cleaner if `init_cutensornet` is a context manager outside `MPS` that returns the handle and `simulate` and the MPS inits can receive it as an optional argument. I'll write a short PR tomorrow morning doing that.

Should I merge this first as it is? or would it be best to merge the change of `init_cuntensornet` first?

Reply:
Aha, sounds good! Up to you really, I don't mind either way. If you first merge this one, then pls. adapt the example in the next PR!

Reply:
I will merge this PR first then, since the other PR will need to change the use of `init_cutensornet` everywhere else, including the Jupyter Notebook and tests.